### PR TITLE
Updated the retry rule

### PR DESCRIPTION
**Pull request by Gannalyo**
1.The original rule
	1.1 retries > 0, it will retry the given times at most
	1.2 retries = 0, it will execute only once
	1.3 retries = -1, it will retry forever until interuppted

2.Reason
	2.1 retries < -1, it will minus one each retry, it will become to be the max value of Java Integer when the minimum value of Java Integer minus one, so the vlaue can be zero, then it has to stop the retry.
	2.2 Hope to wipe off the special value '-1', just from product perspective. So that, it has such a legible logic.

3.The latest rule
	3.1 retries > 0, it will retry the given times at most
	3.2 retries = 0, it will execute only once
	3.3 retries < 0, it will retry forever until interuppted

修改重试逻辑规则

1.原逻辑
	1.1 retries > 0时，最多重试 retries 次
	1.2 retries = 0时，不重试
	1.3 retries = -1时，无限重试

2.修改原因
	2.1 若retries < -1，则每次重试都会减1，当值减到Integer的最小值（-2147483647）时，依据Java中Integer数据类型的计算规则，再减1则为Integer的最大值2147483647，继续重试，当值减到0则会终止重试
	2.2 站在产品角度考虑，建议去除特殊值-1，仅保留大于0，等于0，小于0三种场景，逻辑清晰易懂

3.现逻辑
	3.1 retries > 0时，重试 retries 次
	3.2 retries = 0时，不重试
	3.3 retries < 0时，无限重试

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by zhfeng** [comment]
@Gannalyo Thanks for your contribution and can you create a JIRA on https://issues.apache.org/jira/projects/SCB to track the issue ?

File: omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/ForwardRecovery.java [hunk 1]
```diff
@@ -52,8 +53,9 @@ public Object apply(ProceedingJoinPoint joinPoint, Compensable compensable, Comp
             throw throwable;
           }
 
-          remains = remains == -1 ? -1 : remains - 1;
-          if (remains == 0) {
+          if (remains > 0) {
+            remains--;
+          } else if (remains == 0 ) {
```
Comment (WillemJiang) [inline omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/ForwardRecovery.java]: The default version of remains is 0.  we will always get the error log for the default value.

Comment (Gannalyo) [reply omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/ForwardRecovery.java]: 不知道您是不是没有看懂我那很烂的英文。。。所以我写中文

老程序重试值为-1是一直重试，但如果设置重试值为“-2147483647“，老程序还会继续减减，“-2147483647”再减1就变为了正数的“2147483647”，而正数“2147483647”还会继续减减，当减到0时，就停止重试了。

Comment (WillemJiang) [reply omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/ForwardRecovery.java]: 目前remains 的设置要么大于0， 要么只能是-1， 小于-1的情况应该是要抛出异常错误的。

Comment (Gannalyo) [reply omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/ForwardRecovery.java]: 希望你们仔细看下此处程序逻辑，我们这面测试测出来的问题。

Comment (WillemJiang) [reply omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/ForwardRecovery.java]: 现在我们修复这个问题需要做的是当remains小于-1的时候抛异常，而不是把参数范围修改成支持小于-1的设置。还有就是之前我已经说了你的PR有一些副作用，导致单元测试失败。 我在review日志里面已经提到了。

**Comment by WillemJiang** [comment]
I just fill a [JIRA](https://issues.apache.org/jira/browse/SCB-1036) for the issue.
I will commit a quick fix for it shortly. Do you mind close the issue once you verify the patch.

**Comment by Gannalyo** [comment]
您的意思是如果设置retries < -1，启动时将报错，是吧？？

恕我直言，您为什么要保留-1这个特殊值呢？？后续在对外讲述重试逻辑时，很麻烦呀，包含了四种（>0、0、-1、<-1）后续使用者很容易混淆呀。为什么不正数、0、负数呢？这很好理解呀。还有设置<-1时，程序还无法启动，感觉这两处很不合理。

感觉增加了使用复杂度，如果我是使用者，随便我设置什么值，程序都能正常运行，且逻辑上越简单越好。

强烈建议去除-1特殊值，仅保留正数、0和负数的重试逻辑。

**Comment by WillemJiang** [comment]
对现在的逻辑就是如果小于-1就报错，不让用户设置小于-1的retries值.

**Comment by zhfeng** [comment]
@Gannalyo can you verify the lastest release of the servicecomb-pack ? I will close this PR since it looks like some conflicts.

**Pull request closed without merging**